pkg/scheduler/objects/events.go

```diff
@@ -50,6 +51,71 @@ func (evt *applicationEvents) sendPlaceholderLargerEvent(ph *Allocation, request
 	evt.eventSystem.AddEvent(event)
 }
 
+func (evt *applicationEvents) sendNewAllocationEvent(alloc *Allocation) {
+	if !evt.enabled {
+		return
+	}
+
+	event := events.CreateAppEventRecord(evt.app.ApplicationID, "", alloc.GetUUID(), si.EventRecord_ADD, si.EventRecord_APP_ALLOC, alloc.GetAllocatedResource())
+	evt.eventSystem.AddEvent(event)
+}
+
+func (evt *applicationEvents) sendNewAskEvent(request *AllocationAsk) {
+	if !evt.enabled {
+		return
+	}
+
+	event := events.CreateAppEventRecord(evt.app.ApplicationID, "", request.GetAllocationKey(), si.EventRecord_ADD, si.EventRecord_APP_REQUEST, request.GetAllocatedResource())
+	evt.eventSystem.AddEvent(event)
+}
+
+func (evt *applicationEvents) sendRemoveAllocationEvent(alloc *Allocation, terminationType si.TerminationType) {
+	if !evt.enabled {
+		return
+	}
+
+	var event *si.EventRecord
+	switch terminationType {
+	case si.TerminationType_UNKNOWN_TERMINATION_TYPE:
+		event = events.CreateAppEventRecord(evt.app.ApplicationID, "", alloc.GetUUID(), si.EventRecord_REMOVE, si.EventRecord_ALLOC_NODEREMOVED, alloc.GetAllocatedResource())
+	case si.TerminationType_STOPPED_BY_RM:
+		event = events.CreateAppEventRecord(evt.app.ApplicationID, "", alloc.GetUUID(), si.EventRecord_REMOVE, si.EventRecord_ALLOC_CANCEL, alloc.GetAllocatedResource())
+	case si.TerminationType_TIMEOUT:
+		event = events.CreateAppEventRecord(evt.app.ApplicationID, "", alloc.GetUUID(), si.EventRecord_REMOVE, si.EventRecord_ALLOC_TIMEOUT, alloc.GetAllocatedResource())
+	case si.TerminationType_PREEMPTED_BY_SCHEDULER:
+		event = events.CreateAppEventRecord(evt.app.ApplicationID, "", alloc.GetUUID(), si.EventRecord_REMOVE, si.EventRecord_ALLOC_PREEMPT, alloc.GetAllocatedResource())
+	case si.TerminationType_PLACEHOLDER_REPLACED:
+		event = events.CreateAppEventRecord(evt.app.ApplicationID, "", alloc.GetUUID(), si.EventRecord_REMOVE, si.EventRecord_ALLOC_REPLACED, alloc.GetAllocatedResource())
+	}
+
+	evt.eventSystem.AddEvent(event)
+}
+
+func (evt *applicationEvents) sendRemoveAskEvent(request *AllocationAsk, terminationType si.TerminationType) {
```

Comment:
It's ok to have a single method for remove/cancel, just merge the two.

Reply:
It looks like we don't have a termination type distinguish between "application removal" and "request removal". Currently, I use `TerminationType_STOPPED_BY_RM` for "request removal". Can I use `TerminationType_UNKNOWN_TERMINATION_TYPE` for "application removal"?

Reply:
You shouldn't be concerned with application removal, that's a different JIRA ticket. This task is only about allocations/asks.

Reply:
Updated it. Thanks!